Makefile

```diff
@@ -22,17 +37,98 @@ help:
 	@echo "                                      V E C T O R"
 	@echo ""
 	@echo "---------------------------------------------------------------------------------------"
+	@echo -e "Nix user? You can use \033[0;33m\`direnv allow .\`\033[0m or \033[0;33m\`nix-shell --pure\`\033[0m"
+	@echo -e "Want to use \033[0;33m\`docker\`\033[0m or \033[0;33m\`podman\`\033[0m? See \033[0;33m\`ENVIRONMENT=true\`\033[0m commands. (Default \033[0;33m\`CONTAINER_TOOL=docker\`\033[0m)"
 	@echo ""
 	@awk 'BEGIN {FS = ":.*##"; printf "Usage: make \033[36m<target>\033[0m\n"} /^[a-zA-Z0-9_-]+:.*?##/ { printf "  \033[36m%-46s\033[0m %s\n", $$1, $$2 } /^##@/ { printf "\n\033[1m%s\033[0m\n", substr($$0, 5) } ' $(MAKEFILE_LIST)
 
-##@ Default
+##@ Environment
+# We use a volume here as non-Linux hosts are extremely slow to share disks, and Linux hosts tend to get permissions clobbered.
+define ENVIRONMENT_EXEC
+	@echo "Entering environment..."
+	@mkdir -p target
+	$(CONTAINER_TOOL) run \
+			--name vector-environment \
+			--rm \
+			$(if $(findstring true,$(ENVIRONMENT_TTY)),--tty,) \
+			--init \
+			--interactive \
+			--env INSIDE_ENVIRONMENT=true \
+			--network host \
+			--mount type=bind,source=${PWD},target=/vector \
+			--mount type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock \
+			--mount type=volume,source=vector-target,target=/vector/target \
+			--mount type=volume,source=vector-cargo-cache,target=/root/.cargo \
+			$(ENVIRONMENT_UPSTREAM)
+endef
+
+define ENVIRONMENT_COPY_ARTIFACTS
+	@echo "Copying artifacts off volumes... (Docker errors below are totally okay)"
+	@mkdir -p ./target/release
+	@mkdir -p ./target/debug
+	@mkdir -p ./target/criterion
+	@$(CONTAINER_TOOL) rm -f vector-build-outputs || true
+	@$(CONTAINER_TOOL) run \
+		-d \
+		-v vector-target:/target \
+		--name vector-build-outputs \
+		busybox true
+	@$(CONTAINER_TOOL) cp vector-build-outputs:/target/release/vector ./target/release/ || true
+	@$(CONTAINER_TOOL) cp vector-build-outputs:/target/debug/vector ./target/debug/ || true
+	@$(CONTAINER_TOOL) cp vector-build-outputs:/target/criterion ./target/criterion || true
+	@$(CONTAINER_TOOL) rm -f vector-build-outputs
+endef
+
+
+ifeq ($(ENVIRONMENT_AUTOBUILD), true)
```

Comment:
Can we add an option to skip `ENVIRONMENT_PREPARE` altogether? It adds a tiny bit of time to each invocation, and it would be great to have an option to skip it sometimes.

Reply:
I suggest you use Nix in this case. :)